### ISSUE TITLE
Assertion failed: (ret != SQL_INVALID_HANDLE), function GetColumnValue, file ../src/odbc.cpp, line 576. 

### ISSUE THREAD

**Issue by gopi014** [ibmdb/node-ibm_db#293]
Hi ,
I am using ibm_db node module to connect to my server db2 database. The Db2 is running on a linux machine. Very frequently I am getting the above error and my app gets crashed immediately because of which my users are thrown back to login page. Can you please help me as how I can handle the above error so that my app will catch this error instead app getting crashed.  I am using bluebird for making the connections.

this is my db-config file
module.exports = {
	DATABASE: 'mydb',
	HOSTNAME: 'mydbhost',
	UID: 'myusername',
	PWD: 'mypassword',
	PORT: 60000,
	PROTOCOL: 'TCPIP'
}
this is my ibmdb file 
const Promise = require('bluebird');
const ibmdb = Promise.promisifyAll(require('ibm_db'));
const pool = new ibmdb.Pool();
const config = require('../db/db-config');

module.exports = {
	open() {
		return pool.openAsync(config);
	},

	prepareAndExecute(conn, sql, bindingParameters) {
		console.log("H SQL :- "+sql+" : "+bindingParameters);
		return conn.prepareAsync(sql)
			.then(stmt => stmt.executeAsync(bindingParameters))
			.then(result => {
				const data = result.fetchAllSync();
				result.closeSync();
				return data;
			});
	},

	prepareAndExecuteNonQuery(conn, sql, bindingParameters) {
		console.log("H Insert SQL :- "+sql+" : "+bindingParameters);
		return conn.prepareAsync(sql)
			.then(stmt => stmt.executeNonQueryAsync(bindingParameters));
	}
};

and this is how I make my db calls 
`const ibmdb = require('../db/ibmdb');
return new Promise(function(resolve, reject) {
  const sql = "SELECT * from mytable ";
  ibmdb.open().then(conn => {
    return ibmdb.prepareAndExecute(conn, sql)
      .catch(err => {
        conn.closeSync();
        console.log(err.message);
        throw err;
      })
      .then(data => {
        conn.closeSync();
        resolve(data);
      });
  });

});`

**Comment by rhtpandeyIN** [comment]
Could you be able to post the full error here? Thanks

**Comment by gopi014** [comment]
it just gives this error Assertion failed: (ret != SQL_INVALID_HANDLE), function GetColumnValue, file ../src/odbc.cpp, line 576. and then my node js app gets crashed post that .

**Comment by rhtpandeyIN** [comment]
What is the specific scenarios and conditions when this error is coming? do you have any repro. steps?
If possible can you share the application debug log too.

**Comment by gopi014** [comment]
there is no specific scenarios .when i execute the same query 90 out 100 times it passes and 10 times it fails and this is what i see in the log

Assertion failed: (ret != SQL_INVALID_HANDLE), function GetColumnValue, file ../src/odbc.cpp, line 576.
[nodemon] app crashed - waiting for file changes before starting...

other than this no other errors are coming up

**Comment by rhtpandeyIN** [comment]
Hi, I remember the same case we got few days ago, one of our customer was facing the same issue and after checking their application debug log we found that - it was the application logic error, not because of node-ibm_db.

The possible explanation of this particular problem is : 

`
Well I checked the provided application log and it looks like that - might be application logic is not written properly (at least not working as expected - calling api sync problem),

I observed that might be - some calling api explicitly closing the connection or statement handle even before calling the "query" api, 
as at the time of "selection query" api internally calls "GetColumnValue" and  "GetColumnValue" uses the statement handle to execute further. 
I think, here statement handle is getting closed while query is still returning data, that's why "Invalid Handle" error is coming, which is quite expected and right.

(may be calling api's are not in sync)

2017-07-06T10:28:29.57+0900 [APP/PROC/WEB/0]OUT Invalid Handle: SQLGetData retrun code = -2, stmt handle = 3:4, columnType = -5, index = 8
2017-07-06T10:28:29.57+0900 [APP/PROC/WEB/0]ERR node: ../src/odbc.cpp:609: static v8::Handle<v8::Value> ODBC::GetColumnValue(SQLHSTMT, Column, uint16_t*, int): Assertion `ret != -2' failed.
`
